File: typescript/game.ts
import * as readline from 'readline';

interface IPlayer {
    name: string;
    health: number;
    score: number;
    level: number;
    x: number;
    y: number;
}

class Player implements IPlayer {
    name: string;
    health: number = 100;
    score: number = 0;
    level: number = 1;
    x: number = 0;
    y: number = 0;

    constructor(name: string) {
        this.name = name || 'Hero';
    }

    move(dx: number, dy: number): void {
        this.x += dx;
        this.y += dy;
    }

    takeDamage(amount: number): boolean {
        this.health = Math.max(0, this.health - amount);
        return this.health > 0;
    }

    addScore(points: number): void {
        this.score += points;
    }

    heal(): void {
        if (this.health < 100) {
            const healAmount = Math.min(20, 100 - this.health);
            this.health += healAmount;
            console.log(`Healed for ${healAmount} HP`);
        } else {
            console.log('Already at full health!');
        }
    }

    displayStatus(): void {
        console.log(`\n[${this.name}] HP: ${this.health} | Score: ${this.score} | Level: ${this.level}`);
        console.log(`Position: (${this.x}, ${this.y})`);
    }
}

class GameEngine {
    private player: Player;
    private running: boolean = true;
    private rl: readline.Interface;

    constructor(player: Player) {
        this.player = player;
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
    }

    async run(): Promise<void> {
        console.log('\n=== TypeScript Game Engine ===');
        console.log('Welcome to the TypeScript Game!\n');

        while (this.running && this.player.health > 0) {
            this.player.displayStatus();
            await this.handleInput();
            this.updateGame();
        }

        this.endGame();
        this.rl.close();
    }

    private handleInput(): Promise<void> {
        return new Promise((resolve) => {
            this.rl.question('Commands: (m)ove, (a)ttack, (h)eal, (q)uit: ', (input: string) => {
                const command = input.toLowerCase().trim()[0];
                switch (command) {
                    case 'm':
                        const dx = Math.floor(Math.random() * 5) - 2;
                        const dy = Math.floor(Math.random() * 5) - 2;
                        this.player.move(dx, dy);
                        console.log(`Moved to (${this.player.x}, ${this.player.y})`);
                        break;
                    case 'a':
                        this.attack();
                        break;
                    case 'h':
                        this.player.heal();
                        break;
                    case 'q':
                        this.running = false;
                        break;
                    default:
                        console.log('Unknown command');
                }
                resolve();
            });
        });
    }

    private attack(): void {
        const damage = Math.floor(Math.random() * 21) + 10;
        console.log(`You attacked for ${damage} damage!`);
        this.player.addScore(damage);
    }

    private updateGame(): void {
        if (Math.random() < 0.15) {
            const damage = Math.floor(Math.random() * 11) + 5;
            console.log(`\nEnemy attacked for ${damage} damage!`);
            this.player.takeDamage(damage);
        }
    }

    private endGame(): void {
        console.log('\n=== GAME OVER ===');
        console.log(`Final Score: ${this.player.score}`);
    }
}

async function main(): Promise<void> {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    rl.question('Enter your player name: ', (name: string) => {
        rl.close();
        const player = new Player(name);
        const engine = new GameEngine(player);
        engine.run();
    });
}

main().catch(console.error);

File: csharp/Game.cs
using System;
using System.Collections.Generic;

public class Player
{
    public string Name { get; set; }
    public int Health { get; set; }
    public int Score { get; set; }
    public int Level { get; set; }
    public int X { get; set; }
    public int Y { get; set; }

    public Player(string name)
    {
        Name = name ?? "Hero";
        Health = 100;
        Score = 0;
        Level = 1;
        X = 0;
        Y = 0;
    }

    public void Move(int dx, int dy)
    {
        X += dx;
        Y += dy;
    }

    public bool TakeDamage(int amount)
    {
        Health = Math.Max(0, Health - amount);
        return Health > 0;
    }

    public void AddScore(int points)
    {
        Score += points;
    }

    public void Heal()
    {
        if (Health < 100)
        {
            int healAmount = Math.Min(20, 100 - Health);
            Health += healAmount;
            Console.WriteLine($"Healed for {healAmount} HP");
        }
        else
        {
            Console.WriteLine("Already at full health!");
        }
    }

    public void DisplayStatus()
    {
        Console.WriteLine($"\n[{Name}] HP: {Health} | Score: {Score} | Level: {Level}");
        Console.WriteLine($"Position: ({X}, {Y})");
    }
}

public class GameEngine
{
    private Player player;
    private bool running;
    private Random random;

    public GameEngine(Player player)
    {
        this.player = player;
        this.running = true;
        this.random = new Random();
    }

    public void Run()
    {
        Console.WriteLine("\n=== C# Game Engine ===");
        Console.WriteLine("Welcome to the C# Game!\n");

        while (running && player.Health > 0)
        {
            player.DisplayStatus();
            HandleInput();
            UpdateGame();
        }

        EndGame();
    }

    private void HandleInput()
    {
        Console.Write("Commands: (m)ove, (a)ttack, (h)eal, (q)uit: ");
        string input = Console.ReadLine()?.ToLower() ?? "";

        if (!string.IsNullOrEmpty(input))
        {
            switch (input[0])
            {
                case 'm':
                    int dx = random.Next(-2, 3);
                    int dy = random.Next(-2, 3);
                    player.Move(dx, dy);
                    Console.WriteLine($"Moved to ({player.X}, {player.Y})");
                    break;
                case 'a':
                    Attack();
                    break;
                case 'h':
                    player.Heal();
                    break;
                case 'q':
                    running = false;
                    break;
                default:
                    Console.WriteLine("Unknown command");
                    break;
            }
        }
    }

    private void Attack()
    {
        int damage = random.Next(10, 31);
        Console.WriteLine($"You attacked for {damage} damage!");
        player.AddScore(damage);
    }

    private void UpdateGame()
    {
        if (random.NextDouble() < 0.15)
        {
            int damage = random.Next(5, 16);
            Console.WriteLine($"\nEnemy attacked for {damage} damage!");
            player.TakeDamage(damage);
        }
    }

    private void EndGame()
    {
        Console.WriteLine("\n=== GAME OVER ===");
        Console.WriteLine($"Final Score: {player.Score}");
    }
}

class Program
{
    static void Main()
    {
        Console.Write("Enter your player name: ");
        string name = Console.ReadLine();

        Player player = new Player(name);
        GameEngine engine = new GameEngine(player);
        engine.Run();
    }
}
